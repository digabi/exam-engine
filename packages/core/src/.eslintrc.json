{
  "rules": {
    "no-restricted-imports": [
      "error",
      {
        "paths": [
          {
            "name": "lodash",
            "message": "Please use import * as _ from 'lodash-es' in the core module for better tree shaking."
          },
          {
            "name": "react-i18next",
            "importNames": ["useTranslation", "Translation"],
            "message": "Please use the `useExamTranslation` wrapper instead."
          }
        ]
      }
    ]
  }
}
