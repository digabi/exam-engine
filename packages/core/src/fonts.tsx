import FontFaceObserver from 'fontfaceobserver'
import * as _ from 'lodash-es'

/** A promise that resolves when the main fonts on the exam have been loaded. */
export const ready = Promise.all([
  new FontFaceObserver('Noto Sans', { weight: 300 }).load(), // Light
  new FontFaceObserver('Noto Sans', { weight: 400 }).load(), // Regular
  new FontFaceObserver('Noto Sans', { weight: 600 }).load(), // Semibold
])
  // eslint-disable-next-line promise/always-return
  .then(() => {
    loaded = true
  })
  .catch(_.noop)

/** A boolean flag that tells if the main fonts have been loaded.  */
export let loaded = false
