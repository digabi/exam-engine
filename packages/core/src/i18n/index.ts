import i18n from 'i18next'
import * as _ from 'lodash-es'
import { A_E } from './A_E'
import { BA } from './BA'
import { CA } from './CA'
import { EA } from './EA'
import { FA } from './FA'
import { fi_FI } from './fi-FI'
import { GC } from './GC'
import { I } from './I'
import { M } from './M'
import { O_E } from './O_E'
import { PA } from './PA'
import { SA } from './SA'
import { sv_FI } from './sv-FI'
import { TC } from './TC'
import { VA } from './VA'
import { Z } from './Z'

const examSpecificTranslations = {
  A_E,
  BA,
  BB: BA,
  CA,
  CB: CA,
  EA,
  EC: EA,
  FA,
  FC: FA,
  GC,
  I,
  M,
  N: M,
  O_E,
  SA,
  SC: SA,
  PA,
  PC: PA,
  TC,
  VA,
  VC: VA,
  Z,
}

const resources = {
  'fi-FI': {
    translation: fi_FI,
    ...examSpecificTranslations,
  },
  'sv-FI': {
    translation: sv_FI,
    ...examSpecificTranslations,
  },
}

export function initI18n(language: string, examCode: string | null, dayCode: string | null): typeof i18n {
  const namespace = examCode ? examCode + (dayCode ? `_${dayCode}` : '') : 'translation'

  return i18n.createInstance(
    {
      resources,
      lng: language,
      fallbackLng: 'fi-FI',
      debug: false,
      defaultNS: namespace,
      fallbackNS: 'translation',
      interpolation: {
        escapeValue: false,
        format: (value, format) => {
          switch (format) {
            case 'range': {
              // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
              const [start, end] = value
              // eslint-disable-next-line @typescript-eslint/no-unsafe-return,@typescript-eslint/restrict-template-expressions
              return start == null ? end : end == null ? start : start === end ? start : `${start}â€“${end}`
            }
            case 'first':
              return _.first(value)
            case 'last':
              return _.last(value)
            default:
              // eslint-disable-next-line @typescript-eslint/no-unsafe-return
              return value
          }
        },
      },
    },
    _.noop
  )
}

export const changeLanguage = (_i18n: typeof i18n, language: string) => (): void => {
  if (_i18n.language !== language) {
    _i18n.changeLanguage(language).then(_.noop).catch(_.noop)
  }
}

export default i18n
