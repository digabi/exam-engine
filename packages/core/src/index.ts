import '../src/css/main.less'
import Attachments from './components/attachments/Attachments'
import Exam from './components/exam/Exam'
import GradingInstructions from './components/grading-instructions/GradingInstructions'
import Results from './components/results/Results'
import parseExam from './parser/parseExam'

export { Attachments, Exam, GradingInstructions, Results, parseExam }
export { ExamComponentProps } from './createRenderChildNodes'
export * from './types/ExamAnswer'
export * from './types/ExamServerAPI'
export * from './types/GradingStructure'
export * from './types/Score'

export interface ExamBundle {
  Attachments: typeof Attachments
  Exam: typeof Exam
  parseExam: typeof parseExam
  Results: typeof Results
}
