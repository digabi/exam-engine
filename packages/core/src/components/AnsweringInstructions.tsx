import * as _ from 'lodash-es'
import React from 'react'
import { useExamTranslation } from '../i18n'
export interface AnsweringInstructionProps {
  maxAnswers: number
  /** Child questions as either raw Elements from the XML or just display numbers. */
  childQuestions: (string | Element)[]
  elementType: 'exam' | 'question' | 'section' | 'toc-section'
  minAnswers?: number
}

function AnsweringInstructions({ maxAnswers, minAnswers, elementType, childQuestions }: AnsweringInstructionProps) {
  const childDisplayNumbers = childQuestions.map((question) =>
    _.isString(question) ? question : question.getAttribute('display-number')!
  )

  const { t } = useExamTranslation()
  const possibleTranslationStrings: any[] = cartesian(
    [childDisplayNumbers.length, '*'],
    [maxAnswers, '*'],
    [minAnswers != null ? minAnswers : maxAnswers, '*']
  ).map(
    ([childQuestionCount, maxAnswerCount, minAnswerCount]) =>
      `answering-instructions.${elementType}_${childQuestionCount}_${minAnswerCount}_${maxAnswerCount}`
  )
  const answerCount = [minAnswers, maxAnswers]
  const questions = [_.first(childDisplayNumbers)!, _.last(childDisplayNumbers)!]
  return <>{t(possibleTranslationStrings, { answerCount, questions })}</>
}

function cartesian<A, B, C>(as: readonly A[], bs: readonly B[], cs: readonly C[]): [A, B, C][] {
  const result: [A, B, C][] = []

  for (const a of as) {
    for (const b of bs) {
      for (const c of cs) {
        result.push([a, b, c])
      }
    }
  }

  return result
}

export default React.memo(AnsweringInstructions)
