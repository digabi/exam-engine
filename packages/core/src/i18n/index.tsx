import i18n, { PostProcessorModule, TOptions } from 'i18next'
import * as _ from 'lodash-es'
import React from 'react'
// eslint-disable-next-line no-restricted-imports
import { TFunction, useTranslation } from 'react-i18next'
import { A_E } from './A_E'
import { BA } from './BA'
import { CA } from './CA'
import { EA } from './EA'
import { FA } from './FA'
import { fi_FI } from './fi-FI'
import { GC } from './GC'
import { I } from './I'
import { M } from './M'
import { O_E } from './O_E'
import { PA } from './PA'
import { SA } from './SA'
import { sv_FI } from './sv-FI'
import { TC } from './TC'
import { VA } from './VA'
import { Z } from './Z'

const examSpecificTranslations = {
  A_E,
  BA,
  BB: BA,
  CA,
  CB: CA,
  EA,
  EC: EA,
  FA,
  FC: FA,
  GC,
  I,
  M,
  N: M,
  O_E,
  SA,
  SC: SA,
  PA,
  PC: PA,
  TC,
  VA,
  VC: VA,
  Z,
}

const resources = {
  'fi-FI': {
    translation: fi_FI,
    ...examSpecificTranslations,
  },
  'sv-FI': {
    translation: sv_FI,
    ...examSpecificTranslations,
  },
}

declare module 'react-i18next' {
  type ExamEngineI18nResources = { translation: typeof fi_FI }

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface Resources extends ExamEngineI18nResources {}
}

export function initI18n(language: string, examCode?: string, dayCode?: string): typeof i18n {
  const namespace = examCode ? examCode + (dayCode ? `_${dayCode}` : '') : 'translation'

  return i18n
    .use<PostProcessorModule>({
      type: 'postProcessor',
      name: 'lowercase',
      process: (value, _key, options) => value.toLocaleLowerCase(options.lng),
    })
    .createInstance(
      {
        compatibilityJSON: 'v3',
        resources,
        lng: language,
        fallbackLng: 'fi-FI',
        debug: false,
        defaultNS: namespace,
        fallbackNS: 'translation',
        interpolation: {
          escapeValue: false,
          format: (value, format) => {
            const val = value as [string, string]
            switch (format) {
              case 'range': {
                const [start, end] = val
                return start == null ? end : end == null ? start : start === end ? start : `${start}–${end}`
              }
              case 'first':
                return _.first(val)
              case 'last':
                return _.last(val)
              default:
                // eslint-disable-next-line @typescript-eslint/no-unsafe-return
                return value
            }
          },
        },
      },
      _.noop
    )
}

export function useExamTranslation() {
  const { i18n, t: originalT } = useTranslation()
  type Translator = {
    (key: string | string[], options?: TOptions): string | JSX.Element
    raw: TFunction
  }
  const t: Translator = (key, options) => translate(i18n, key, options)
  t.raw = originalT
  return { i18n, t }
}

/**
 * Custom wrapper for i18n.t that wraps the translated text in a span element
 * containing the correct lang attribute, if necessary.
 *
 * If you want to disable the wrapping, you can use `t.raw(…)`.
 */
function translate(instance: typeof i18n, key: string | string[], options?: TOptions): string | JSX.Element {
  const translation = instance.t(key, options)

  const namespace = instance.options.defaultNS!.toString()
  const examHasCustomTranslations = namespace !== 'translation' && namespace in examSpecificTranslations
  const hasCustomTranslationForKey = examHasCustomTranslations && hasTranslation(instance, namespace, key)
  return hasCustomTranslationForKey || !examHasCustomTranslations ? (
    translation
  ) : (
    <span lang={instance.language}>{translation}</span>
  )
}

/**
 * A hacky way to check if we have a translation in the specified namespace.
 * Ideally, we'd probably want to use `i18n.exists()` for this purpose, but I
 * couldn't find a way to disable the fallback mechanism when using it.
 */
function hasTranslation(instance: typeof i18n, namespace: string, key: string | string[]): boolean {
  return Array.isArray(key)
    ? key.some((k) => instance.getResource(instance.language, namespace, k) != null)
    : instance.getResource(instance.language, namespace, key) != null
}

export const changeLanguage = (_i18n: typeof i18n, language: string) => (): void => {
  if (_i18n.language !== language) {
    _i18n.changeLanguage(language).then(_.noop).catch(_.noop)
  }
}

export default i18n
