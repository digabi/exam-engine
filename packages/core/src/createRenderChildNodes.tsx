import React from 'react'
import { mapChildNodes } from './dom-utils'

export const ExamNamespaceURI = 'http://ylioppilastutkinto.fi/exam.xsd'
export const XHTMLNamespaceURI = 'http://www.w3.org/1999/xhtml'

export interface ExamComponentProps {
  className?: string
  /** An element in the exam XML. */
  element: Element
  /** A function that knows how to render the child nodes of this element. */
  renderChildNodes: RenderChildNodes
}

export const enum RenderOptions {
  RenderHTML,
  SkipHTML,
}

export type RenderChildNodes = (element: Element, options?: RenderOptions) => React.ReactNode[]

export function createRenderChildNodes(
  componentMap: Record<string, React.ComponentType<ExamComponentProps>>
): RenderChildNodes {
  function renderChildNode(node: ChildNode, index: number, options: RenderOptions): React.ReactNode {
    switch (node.nodeType) {
      case Node.TEXT_NODE:
      case Node.CDATA_SECTION_NODE:
        return options === RenderOptions.RenderHTML ? renderTextNode(node as Text) : null
      case Node.ELEMENT_NODE:
        return renderElement(node as Element, index, options)
      default:
        return null
    }
  }

  function renderElement(element: Element, index: number, options: RenderOptions) {
    switch (element.namespaceURI) {
      case XHTMLNamespaceURI:
        return renderXHTMLElement(element as HTMLElement, index, options)
      case ExamNamespaceURI:
        return renderExamElement(element, index)
      default:
        throw new Error(`Unrecognized element: ${element.localName}`)
    }
  }

  function renderXHTMLElement<T extends HTMLElement>(element: T, index: number, options: RenderOptions) {
    const children = element.hasChildNodes() ? renderChildNodes(element, options) : null

    if (options === RenderOptions.RenderHTML) {
      const Tag = element.localName
      const props = htmlAttributes2props(element, index)
      return <Tag {...props}>{children}</Tag>
    } else {
      return children
    }
  }

  function renderExamElement(element: Element, index: number) {
    const Component = componentMap[element.localName]
    const className = element.getAttribute('class') || undefined
    return Component ? <Component {...{ element, className, renderChildNodes, key: key(element, index) }} /> : null
  }

  function renderChildNodes(element: Element, options: RenderOptions = RenderOptions.RenderHTML): React.ReactNode[] {
    return mapChildNodes(element, (childElement, i) => renderChildNode(childElement, i, options))
  }

  return renderChildNodes
}

function renderTextNode(node: Text) {
  return node.textContent!
}

function htmlAttributes2props<T extends HTMLElement>(element: T, index: number): React.HTMLProps<T> {
  const props: React.HTMLProps<T> = { key: key(element, index) }
  for (const attribute of element.attributes) {
    const name = attribute.localName === 'class' ? 'className' : (attribute.localName as keyof React.HTMLProps<T>)
    props[name] = attribute.value
  }
  return props
}

function key(element: Element, index: number): string {
  return `${element.localName}-${index}`
}
