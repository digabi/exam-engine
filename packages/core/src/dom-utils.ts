import * as _ from 'lodash-es'
import { ExamNamespaceURI } from './createRenderChildNodes'
import { useContext, useEffect, useState } from 'react'
import { TOCContext } from './components/context/TOCContext'

export const NBSP = '\u00A0'

export function parentElements(element: Element, selector?: Selector): Element[] {
  const predicate = selector != null ? mkPredicate(selector) : undefined
  const result = []
  while (element.parentElement != null) {
    element = element.parentElement
    if (predicate == null || predicate(element)) {
      result.push(element)
    }
  }
  return result
}

type Selector = string | string[] | ((element: Element) => boolean)

function mkPredicate(selector: Selector): (element: Element) => boolean {
  return typeof selector === 'function'
    ? (element: Element) => element.namespaceURI === ExamNamespaceURI && selector(element)
    : Array.isArray(selector)
      ? (element: Element) => element.namespaceURI === ExamNamespaceURI && selector.includes(element.localName)
      : (element: Element) => element.namespaceURI === ExamNamespaceURI && element.localName === selector
}

export function query(root: Element, selector: Selector): Element | undefined {
  const predicate = mkPredicate(selector)

  function go(element: Element): Element | undefined {
    for (let i = 0, length = element.children.length; i < length; i++) {
      const childElement = element.children[i]
      if (predicate(childElement)) {
        return childElement
      }
      const maybeElement = go(childElement)
      if (maybeElement != null) {
        return maybeElement
      }
    }
  }

  return go(root)
}

export function queryAll(root: Element, selector: Selector, recurse = true): Element[] {
  const predicate = mkPredicate(selector)
  const results: Element[] = []

  function go(element: Element) {
    for (let i = 0, length = element.children.length; i < length; i++) {
      const childElement = element.children[i]
      if (predicate(childElement)) {
        results.push(childElement)
        if (recurse) {
          go(childElement)
        }
      } else {
        go(childElement)
      }
    }
  }

  go(root)
  return results
}

export function queryAncestors(element: Element, selector: Selector): Element | undefined {
  const predicate = mkPredicate(selector)
  while (element.parentElement != null) {
    element = element.parentElement
    if (predicate(element)) {
      return element
    }
  }
}

export function findChildElement(element: Element, selector: Selector): Element | undefined {
  const predicate = mkPredicate(selector)

  for (let i = 0, length = element.children.length; i < length; i++) {
    const childElement = element.children[i]
    if (predicate(childElement)) {
      return childElement
    }
  }
}
export function getAttribute(element: Element, attributeName: string): string | undefined {
  return element.getAttribute(attributeName) ?? undefined
}

export function getNumericAttribute(element: Element, attributeName: string): number | undefined {
  const maybeValue = element.getAttribute(attributeName)
  return maybeValue != null ? Number(maybeValue) : undefined
}

export function getBooleanAttribute(element: Element, attributeName: string): boolean {
  return getAttribute(element, attributeName) === 'true'
}

export function mapChildNodes<T>(element: Element, fn: (childNode: ChildNode, index: number) => T): T[] {
  const length = element.childNodes.length
  const result = new Array<T>(length)
  for (let i = 0; i < length; i++) {
    result[i] = fn(element.childNodes[i], i)
  }
  return result
}

export function mapChildElements<T>(element: Element, fn: (childElement: Element, index: number) => T): T[] {
  const result = []
  for (let i = 0, length = element.children.length; i < length; i++) {
    const childElement = element.children[i]
    result.push(fn(childElement, i))
  }
  return result
}

export function findChildrenAnswers(element: Element): Element[] {
  return queryAll(element, ['choice-answer', 'dropdown-answer', 'text-answer', 'scored-text-answer'], true)
}

export function hasSiblingQuestions(answer: Element) {
  const parentQuestion = queryAncestors(answer, 'question')
  const siblingQuestions = parentQuestion?.querySelectorAll('[question-id]')
  return siblingQuestions && siblingQuestions?.length > 1
}

export function createElement<K extends keyof HTMLElementTagNameMap>(
  tagName: K,
  properties?: Record<string, string> | null,
  ...children: Array<string | ChildNode>
): HTMLElementTagNameMap[K] {
  const element = document.createElement(tagName)
  setProperties(element, properties)
  children.forEach(c => element.appendChild(typeof c === 'string' ? new Text(c) : c))
  return element
}

export function setProperties(element: HTMLElement, properties?: Record<string, string> | null): void {
  _.forEach(properties, (v, k) => {
    if (k === 'css') {
      element.style.cssText = v
    } else if (k.startsWith('data-')) {
      element.setAttribute(k, v)
    } else {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      ;(element as any)[k] = v
    }
  })
}

export function useIsElementInViewport(elementType: 'question' | 'section', displayNumber: string) {
  const { visibleTOCElements } = useContext(TOCContext)
  const [isVisible, setIsVisible] = useState(false)

  const id = `${elementType}-${displayNumber}`

  useEffect(() => {
    if (visibleTOCElements?.includes(id) && !isVisible) {
      setIsVisible(true)
    } else if (!visibleTOCElements?.includes(id) && isVisible) {
      setIsVisible(false)
    }
  }, [visibleTOCElements])

  return isVisible
}

export function getElementPath(element: Element): string {
  const temp = !element.tagName && Array.from(element.parentNode?.childNodes || []).indexOf(element)
  const elementIndex = Array.from(element.parentElement?.children || []).indexOf(element)
  let path = `${element.nodeName}:${elementIndex > 0 ? elementIndex : temp}`
  let currentElement = element
  while (currentElement.parentElement) {
    currentElement = currentElement.parentElement
    const displayNumber = currentElement.getAttribute('display-number')
    const index =
      Number(displayNumber) ||
      Array.from(currentElement.parentElement?.children || [currentElement]).indexOf(currentElement)
    const elementId = currentElement.getAttribute('id')
    path = `${elementId || currentElement.nodeName}:${index} > ${path}`
  }

  return path
}
