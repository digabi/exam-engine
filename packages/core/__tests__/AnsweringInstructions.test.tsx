import * as i18next from 'i18next'
import _ from 'lodash'
import React from 'react'
import { I18nextProvider } from 'react-i18next'
import TestRenderer from 'react-test-renderer'
import AnsweringInstructions, { AnsweringInstructionProps } from '../src/components/AnsweringInstructions'
import { initI18n } from '../src/i18n'

describe('<AnsweringInstructions />', () => {
  let i18n: i18next.i18n

  beforeAll(() => {
    i18n = initI18n('fi-FI')
  })

  describe('question', () => {
    it('numerals', () => {
      expectRendering({
        maxAnswers: 1,
        type: 'question',
        childQuestions: mkQuestions(range(1, 10, '1.')),
      })

      expectRendering({
        maxAnswers: 5,
        type: 'question',
        childQuestions: mkQuestions(range(1, 10, '1.')),
      })
    })

    it('2 child questions and maxAnswers = 1', () => {
      expectRendering({
        maxAnswers: 1,
        type: 'question',
        childQuestions: mkQuestions(range(1, 2, '1.')),
      })
    })

    it('fallback', () => {
      expectRendering({
        maxAnswers: 11,
        type: 'question',
        childQuestions: mkQuestions(range(1, 12, '1.')),
      })
    })
  })

  describe('section', () => {
    it('numerals', () => {
      expectRendering({
        maxAnswers: 1,
        type: 'section',
        childQuestions: mkQuestions(range(1, 10)),
      })

      expectRendering({
        maxAnswers: 5,
        type: 'section',
        childQuestions: mkQuestions(range(1, 10)),
      })
    })

    it('2 child questions and maxAnswers = 1', () => {
      expectRendering({
        maxAnswers: 1,
        type: 'section',
        childQuestions: mkQuestions(range(1, 2)),
      })
    })

    it('fallback', () => {
      expectRendering({
        minAnswers: 2,
        maxAnswers: 5,
        type: 'section',
        childQuestions: mkQuestions(range(1, 12)),
      })

      expectRendering({
        maxAnswers: 11,
        type: 'section',
        childQuestions: mkQuestions(range(1, 12)),
      })
    })
  })

  function range(start: number, end: number, prefix = ''): string[] {
    return _.range(start, end + 1).map((n) => `${prefix}${n}`)
  }

  function expectRendering(props: AnsweringInstructionProps) {
    const container = TestRenderer.create(
      <I18nextProvider i18n={i18n}>
        <AnsweringInstructions {...props} />
      </I18nextProvider>
    )
    expect(container.toJSON()).toMatchSnapshot()
  }

  function mkQuestions(displayNumbers: string[]): Element[] {
    return displayNumbers.map((displayNumber) => {
      const question = document.createElement('question')
      question.setAttribute('display-number', displayNumber)
      return question
    })
  }
})
