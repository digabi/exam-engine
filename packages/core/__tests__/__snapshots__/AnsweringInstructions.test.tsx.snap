// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<AnsweringInstructions /> question 2 child questions and maxAnswers = 1 1`] = `"Vastaa joko kohtaan 1.1 tai 1.2."`;

exports[`<AnsweringInstructions /> question fallback 1`] = `"Vastaa 11 kohdista 1.1–1.12."`;

exports[`<AnsweringInstructions /> question numerals 1`] = `"Vastaa yhteen kohdista 1.1–1.10."`;

exports[`<AnsweringInstructions /> question numerals 2`] = `"Vastaa viiteen kohdista 1.1–1.10."`;

exports[`<AnsweringInstructions /> section 2 child questions and maxAnswers = 1 1`] = `"Vastaa joko tehtävään 1 tai 2."`;

exports[`<AnsweringInstructions /> section fallback 1`] = `"Vastaa 2–5 tehtävään."`;

exports[`<AnsweringInstructions /> section fallback 2`] = `"Vastaa 11 tehtävään."`;

exports[`<AnsweringInstructions /> section numerals 1`] = `"Vastaa yhteen tehtävään."`;

exports[`<AnsweringInstructions /> section numerals 2`] = `"Vastaa viiteen tehtävään."`;
